javascript/ql/lib/semmle/javascript/frameworks/data/internal/ApiGraphModels.qll

```diff
@@ -368,6 +386,58 @@ private predicate invocationMatchesCallSiteFilter(Specific::InvokeNode invoke, A
   Specific::invocationMatchesExtraCallSiteFilter(invoke, token)
 }
 
+private class TypeModelUseEntry extends API::EntryPoint {
+  private string package;
+  private string type;
+
+  TypeModelUseEntry() {
+    exists(any(TypeModel tm).getASource(package, type)) and
+    this = "TypeModelUseEntry;" + package + ";" + type
+  }
+
+  override DataFlow::LocalSourceNode getASource() {
+    result = any(TypeModel tm).getASource(package, type)
+  }
+
+  API::Node getNodeForType(string package_, string type_) {
+    package = package_ and type = type_ and result = this.getANode()
+  }
+}
+
+private class TypeModelDefEntry extends API::EntryPoint {
+  private string package;
+  private string type;
+
+  TypeModelDefEntry() {
+    exists(any(TypeModel tm).getASink(package, type)) and
+    this = "TypeModelDefEntry;" + package + ";" + type
+  }
+
+  override DataFlow::Node getASink() { result = any(TypeModel tm).getASink(package, type) }
+
+  API::Node getNodeForType(string package_, string type_) {
+    package = package_ and type = type_ and result = this.getANode()
+  }
+}
+
+/**
+ * Gets an API node identified by the given `(package,type)` pair.
+ */
+pragma[nomagic]
+private API::Node getNodeFromType(string package, string type) {
+  exists(string package2, string type2, AccessPath path2 |
+    typeModel(package, type, package2, type2, path2) and
+    result = getNodeFromPath(package2, type2, path2, path2.getNumToken())
+  )
+  or
+  result = any(TypeModelUseEntry e).getNodeForType(package, type)
+  or
+  result = any(TypeModelDefEntry e).getNodeForType(package, type)
+  or
+  isRelevantFullPath(package, type, _) and
+  result = Specific::getExtraNodeFromPath(package, type, _, 0)
```

Comment:
I wonder if this should be like:
```suggestion
  exists(string path | isRelevantFullPath(package, type, path) and
  result = Specific::getExtraNodeFromPath(package, type, path, 0))
```
Also these lines overlap with 
```
result = Specific::getExtraNodeFromPath(package, type, path, n)
```
on line [385](https://github.com/github/codeql/pull/10276/files#diff-f6a72b8d40fb2a6a2b8b273c53c19086292a8efa538591bcfb41204f07a327a8L385)

Reply:
Using `getExtraNodeFromPath` with `n=0` was a bit of a hack, actually. The real fix is to introduce `getExtraNodeFromType` instead, PTAL.